### ISSUE TITLE
[FEAT] SentinelHubRequest class ctor has no support for SentinelHubSession

### ISSUE THREAD

**Issue by BlackRider97** [sentinel-hub/sentinelhub-py#146]
**What is the problem? Please describe.**

`SentinelHubRequest`  constructor should accept user defined instance of class `SentinelHubSession` too along with `SHConfig` instance i.e. `config`.  External session should pass till `SentinelHubDownloadClient ` and use user given session for acquiring oAuth token or session headers.

This will enable user to store session externally and re-use across multiple method calls. We are facing problems when we are using SDK in Kubernetes Pods while running lot of parallel jobs. Since there are too many  pods i.e. python processes, authentication token acquire APIs are failing.  

**Here's the solution**
`SentinelHubRequest`  constructor should accept user defined instance of class `SentinelHubSession` too along with `SHConfig` instance i.e. `config`.

**Additional context**
Consider the use cases when SDK is being used like a script in different different processes.

**Comment by AleksMat** [comment]
Hi @BlackRider97,

Thanks for opening and explaining this issue. From your use-case with Kubernetes pods, I see that we definitely need to enable reusing sessions, obtained from external sources.

As you might have seen in the current implementation we are trying to hide sessions from users so that they are taken care of automatically in `SentinelHubDownloadClient`. In order to only use a single session for each given credentials, the `SentinelHubDownloadClient` is caching sessions to a class variable [here](https://github.com/sentinel-hub/sentinelhub-py/blob/master/sentinelhub/download/sentinelhub_client.py#L101).

According to the current implementation, I see 2 options:

- As you suggested we can add a `session` parameter to `SentinelHubRequest` constructor. It turns out it will be a bit tricky to pass this parameter to `SentinelHubDownloadClient` due to the current implementation in a base `DataRequest` class.

- We can add a method to `SentinelHubDownloadClient` to directly cache a new session. The advantage of such method is that the session is given directly to the object which will use it. Because of that it solves the problem for all types of request classes, not only  `SentinelHubRequest`. It also doesn't overload `SentinelHubRequest` constructor with another parameter.

```Python
# First cache a session obtained from an extrenal source
SentinelHubDownloadClient.cache_session(session)

# Make requests without passing a session to it
request = SentinelHubRequest(...)
```
A disadvantage of such approach is that it will be harder for new users to discover it on their own. But I imagine that not many users will need such an advanced handling of sessions.

Let me know any preferences about this.

Another thing that you will have to be careful about is that `SentinelHubSession` object automatically refreshes its token before it would expire. I think a token expires in 1 hour. Before that happens you will have to update the session at one place and pass it to all pods agan. Otherwise, `SentinelHubSession` from each pod will independently from other pods request a new session at about the same time which will again trigger the same problem of having too many open sessions.

**Comment by AleksMat** [comment]
We have just released `sentinelhub-py` version `3.6.0` with a proper support for this. We also prepared an official [tutorial notebook](https://sentinelhub-py.readthedocs.io/en/latest/examples/session_sharing.html) explaining the process and providing different implementations. Any feedback is welcome.
